storage/grpc_client_test.go

```diff
@@ -128,39 +128,67 @@ func TestBytesCodecV2(t *testing.T) {
 		},
 	} {
 		t.Run(test.desc, func(t *testing.T) {
-			// Encode the response.
-			encodedResp, err := proto.Marshal(test.resp)
-			if err != nil {
-				t.Fatalf("proto.Marshal: %v", err)
-			}
-			var respData mem.BufferSlice
-			respData = append(respData, mem.SliceBuffer(encodedResp))
-
-			// Unmarshal and decode response using custom decoding.
-			var encodedBytes *mem.BufferSlice = &mem.BufferSlice{}
-			if err := bytesCodecV2.Unmarshal(bytesCodecV2{}, respData, encodedBytes); err != nil {
-				t.Fatalf("unmarshal: %v", err)
-			}
+			for _, subtest := range []struct {
+				desc        string
+				splitPoints []int
+			}{
+				{
+					desc:        "single buffer",
+					splitPoints: []int{},
+				},
+				{
+					desc:        "random split buffer",
+					splitPoints: []int{100, 500, 700},
+				},
+				{
+					desc:        "mid-tag split buffer",
+					splitPoints: []int{3, 1032},
+				},
+			} {
```

Comment:
I think we should include more tests here, including an empty metadata struct (above). Maybe we could do something like fuzz tests? When I was testing I caught different errors on different split points (mid-tag on different tags, I think maybe end or start of a tag/data as well).

Reply:
It might be interesting to split at every single place possible.  I.e.:

```go
for offs := 1; offs < len(buf); offs++ {
	unmarshal(buf[:offs], buf[offs:])
}
```

And/or a test case that literally puts exactly one byte per buffer.